#!/bin/bash
#PBS -l nodes=1:ppn=16,walltime=12:00:00
#PBS -N viralx4
#PBS -q extended
#PBS -V
#PBS -M sleiman.bassim@stonybrook.edu
#PBS -j oe
#PBS -m abe

set -eux

# __CHANGE__
version=grch
lib=RF

# IACS directories
scratch=/gpfs/scratch/ballam
home=/gpfs/home/ballam
pbs=$(echo $PBS_JOBID | cut -f1 -d '.')

# __DONT CHANGE__
workdir=${scratch}/testing/human
#target=$workdir/human.$transcriptome.selected.500.fa
mkdir -p $workdir
cd ${workdir}

## Map reads to REFERENCE GENOME
mapped2=$workdir/fastq_genome_bwa
_DIR=$workdir/mapped.reads.genome_$version
_DIRt=$workdir/trinity_out_dir_$version
_REF=$(find $scratch/testing/genomes -iname "*$version*fna")
_INDEX=$(find $scratch/testing/genomes -iname "*$version*sa")
mkdir -p $mapped2 $_DIR $_DIRt

## __DONT CHANGE
# index genome
if [ ! -f "$_INDEX" ]; then
    bwa index $_REF
fi

for t in HBV; do
# number of samples
#    for s in {1..2}; do

# Get read files one by one
sense=$(find $scratch/testing/reads -maxdepth 1 -name "$t-R1*gz")
antisense=$(find $scratch/testing/reads -maxdepth 1 -name "$t-R2*gz")

# map and sort by coordinate
bwa mem -t 32 $_REF $sense $antisense | samtools view -Sb -f 0x02 -F12 - | samtools sort -@32 - -o $_DIR/$t.$version.sorted.proper.bam
samtools index $_DIR/$t.$version.sorted.proper.bam

# keep proper paired and discard unmapped reads and mates
for i in 1 2; do
    samtools fastq -$i $mapped2/$t-R$i.fastq $_DIR/$t.$version.sorted.proper.bam
done

 #   done
done

# merge all bams into 1 file for genome guided assembly
samtools merge $_DIR/all.reads.sorted.bam $_DIR/*proper.bam
samtools index $_DIR/all.reads.sorted.bam
#rm $DIR/*proper.bam

$home/trinityrnaseq-2.2.0/Trinity --genome_guided_bam $_DIR/all.reads.sorted.bam --genome_guided_max_intron 1000 --output $_DIRt --CPU 15 --max_memory 50G
