#!/bin/bash
#PBS -l nodes=2:ppn=16,walltime=168:00:00
#PBS -N PARA_GG5
#PBS -q extended
#PBS -V
#PBS -M sleiman.bassim@stonybrook.edu
#PBS -j oe
#PBS -m abe


# __CHANGE__
transcriptome=5
lib=RF

# IACS directories
scratch=/gpfs/scratch/ballam
home=/gpfs/home/ballam
pbs=$(echo $PBS_JOBID | cut -f1 -d '.')

# __DONT CHANGE__
workdir=${scratch}/ganglia/blat/parasite.$transcriptome
target=$workdir/parasite.$transcriptome.selected.500.fa
cd ${workdir}

## Map reads to REFERENCE GENOME
mapped2=$workdir/fastq_genome_bwa
_DIR=$workdir/mapped.reads.genome_trinity
_REF=$(find $scratch/ganglia/genomes/parasite.$transcriptome -name "*fna")
_INDEX=$(find $scratch/ganglia/genomes/parasite.$transcriptome -name "*sa")
mkdir -p $mapped2 $_DIR

## __DONT CHANGE
# index genome
if [ ! -f "$_INDEX" ]; then
    bwa index $_REF
fi

for t in br; do
    for s in {1..2}; do

# Get read files one by one
sense=$(find $scratch/ganglia/raw.reads.trimmed -maxdepth 1 -name "$t.$s.R1.P*gz")
antisense=$(find $scratch/ganglia/raw.reads.trimmed -maxdepth 1 -name "$t.$s.R2.P*gz")

# map and sort by coordinate
bwa mem -t 32 $_REF $sense $antisense | samtools view -Sb -f 0x02 -F12 - | samtools sort -@32 - -o $_DIR/$t.$s.sorted.proper.bam
samtools index $_DIR/$t.$s.sorted.proper.bam

# keep proper paired and discard unmapped reads and mates
for i in 1 2; do
    samtools fastq -$i $mapped2/$t.$s.R$i.fastq $_DIR/$t.$s.sorted.proper.bam
done

    done
done

# merge all bams into 1 file for genome guided assembly
samtools merge $_DIR/all.reads.sorted.bam $_DIR/*proper.bam
samtools index $_DIR/all.reads.sorted.bam
rm $DIR/*proper.bam

$home/trinityrnaseq-2.2.0/Trinity --genome_guided_bam $_DIR/all.reads.sorted.bam --genome_guided_max_intron 1000 --output $_DIR --CPU 15 --max_memory 50G
