#+TITLE: Translational Research: Pipelines in Data Mining and Systems Biology
#+AUTHOR: Sleiman Bassim, PhD
#+EMAIL: slei.bass@gmail.com

#+STARTUP: content
#+STARTUP: hidestars
#+OPTIONS: toc:5 H:5 num:3
#+LANGUAGE: english
#+LaTeX_HEADER: \usepackage[ttscale=.875]{libertine}
#+LATEX_HEADER: \usepackage[T1]{fontenc}
#+LaTeX_HEADER: \sectionfont{\normalfont\scshape}
#+LaTeX_HEADER: \subsectionfont{\normalfont\itshape}
#+LATEX_HEADER: \usepackage[innermargin=1.5cm,outermargin=1.25cm,vmargin=3cm]{geometry}
#+LATEX_HEADER: \linespread{1}
#+LATEX_HEADER: \setlength{\itemsep}{-30pt}
#+LATEX_HEADER: \setlength{\parskip}{0pt}
#+LATEX_HEADER: \setlength{\parsep}{-5pt}
#+LATEX_HEADER: \usepackage[hyperref]{xcolor}
#+LATEX_HEADER: \usepackage[colorlinks=true,urlcolor=SteelBlue4,linkcolor=Firebrick4]{hyperref}
#+EXPORT_SELECT_TAGS: export
#+EXPORT_EXCLUDE_TAGS: noexport

The following repository contains RNA-seq and microarray pipelines for transcriptome
assembly, gene expression, gene annotation, genomic mutation calling, and other. Scripts dated
between 2015 and 2017 are running on [[https://www.xsede.org/][XSEDE]] supercomputers and [[http://www.iacs.stonybrook.edu/resources/handy-accounts#overlay-context=resources/accounts][LIred]] cluster in
the US, including Greenfield, Bridges, and LIred servers. All others
are running on H4H cluster at [[http://www.uhnresearch.ca/][UHN]], in Toronto Canada.

Licence: MIT Licence

* Algorithms will include:
   - Feature engineering & regularization (lasso, ridge, elastic)
   - Data subsetting, extraction, reformatting & report designs
   - Subsampling, mini-batch sampling & bagging
   - Data spliting (binomial and multiclass)
   - Unsupervised learning (fuzzy, hierarchical clustering)
   - Grid search for normalization & standardization methods
   - Bayesian inferential models
   - Similarity & adjacency matrices
   - Multi-iterative module allocations for gene expressions
   - Weighted genetic networks
   - Supervised learning and grid hyper tuning
   - Bootstrapping and model alpha adjustments
   - Logging & performance metrics (ROC, AUROC, 95% CI, kappa)
   - Various descriptive and performance plotting
   - Nested cross-validation & iterative resampling structures
   - Multi-class area under the ROC curve
   - Feature importance scoring
   - Confusion matrices & multi-prediction validation
   - Redundancy and descriptive analyses
   - Machine learning optimizations
   - Random seeding optimizations
   - Over 20 machine learning models
   - Deep learning (Tensor, Torch, Caffe, Mxnet, H2O, Keras)
   - Reinforcement learning (batch, temporal abstraction, deep RL)
   - QC for most next generation sequencing platforms
   - Abundance analysis (FPKM, RPKM) multi-tool cross comparison
   - Genome & targeted exome sequencing
   - Variant calling (SNVs, CNVs, Indels)
   - Germline filtering analysis
   - Gene/Variant annotations (model, non-model species) multi-DB cross matches
   - Differential gene expression (microarrays, RNA-seq) in R/Python


* Systemic Analytical Pipelines
=scripts can be scaled while adjusted for any sample size. Majority
reproducible on other data sets=
1. =hot= [[https://github.com/neocruiser/pipelines/blob/master/r/classification.R][Machine & Deep learning multi-model]] analyses for cancer prediction in R (Bassim 2018)
2. [[https://github.com/neocruiser/pipelines/blob/master/r/line.ranking.R][Gene ranking]] in R & [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.extract.genes.sh][multi-output wrangling]] in bash from transcriptional data (Bassim 2018)
3. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.expression.distribution.R][Polygenic]] & [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.pval.distribution.R][probability distributions]] reported by batch visualization in R (Bassim 2018)
4. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.h4h.pbs][Microarray sample preprocessing & automation]] in bash (Bassim 2018)
5. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.2.0.R][Microarray multi-contrast & batch logging]] in R (Bassim 2018)
6. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.summary.h4h.sh][Aggregated performance metrics]] for transcriptional analyses in bash (Bassim 2018)
7. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.h4h.pbs][Data restructuring & mining]] for network automation in bash (Bassim 2018)
8. =hot= [[Https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.affymetrix.R][Multi-grid search approch with data-driven gene networks]] in R (Bassim 2018)
9. [[https://github.com/neocruiser/pipelines/blob/master/r/heatmaps.3.0.R][Clustering & gene expression with bootstrapped approach & metric aggregation]] in R (Bassim 2018)
10. =hot= [[https://github.com/neocruiser/pipelines/blob/master/calling/exome_calling.v2.pyclone.h4h.pbs][Targeted exome calling]] for variants in bash (Bassim 2018)
11. [[https://github.com/neocruiser/pipelines/blob/master/calling/mpileup.standalone.h4h.pbs][Variant clustering]] for classification of tumor clones in bash and python (Bassim 2018)
12. [[https://github.com/neocruiser/pipelines/blob/master/debug/reinforce_v2.py][Reinforcement]], [[https://github.com/neocruiser/pipelines/blob/master/debug/generative_mod.v1.23.py][Generative CNN]] & [[https://github.com/neocruiser/pipelines/blob/master/debug/deep_v3.py][Deep learning]] for data embedding in python (Bassim 2017)
13. [[https://github.com/neocruiser/pipelines/blob/master/debug/debug2.pbs][Two species reads separation from dual RNA-seq of host & parasite]] in bash (Bassim 2017)
14. [[https://github.com/neocruiser/pipelines/blob/master/debug/debug4.slurm][Sequence abundance and expression from dual RNA-seq]] in R (Bassim 2017)
15. =hot= [[https://github.com/neocruiser/pipelines/blob/master/mining/automated_analyses.sh][Aggregating & optimizing data structure ]]for pathway discovery in bash (Bassim 2016) 
16. [[https://github.com/neocruiser/pipelines/blob/master/expression/filter-bridges.slurm][Reducing transcriptome size]] by correcting for read abundance in bash (Bassim 2017)
17. [[https://github.com/neocruiser/pipelines/blob/master/assembly/trinity-bridges.slurm][Transcriptome assembly of RNA-seq data optimized on high performance clusters]] in bash (Bassim 2016)
18. [[https://github.com/neocruiser/pipelines/blob/master/mapping/genome_guided_assemblies.pbs][Genome-guided assembly of RNA-seq data for new gene discovery]] in bash (Bassim 2016)
19. =hot= [[https://github.com/neocruiser/pipelines/blob/master/expression/degs-bridges.slurm][RNA-seq gene expression multi-approach & metric aggregation]] in bash and R (Bassim 2016)
20. [[https://github.com/neocruiser/pipelines/blob/master/annotation/kraken.db-bridges.slurm][Shotgun sequencing pipeline]] & virus classification or microbes identification in bash (Bassim 2016)
21. [[https://github.com/neocruiser/pipelines/blob/master/annotation/diamond-bridges.slurm][Fast sequence annotation against NCBI databases]] with Blast & Diamond in bash (Bassim 2016)
22. [[https://github.com/neocruiser/pipelines/blob/master/annotation/interproscan-bridges.slurm][Protein sequence annotation & data aggregation]] with multi-database mining in bash (Bassim 2016)
23. [[https://github.com/neocruiser/pipelines/blob/master/annotation/blast-iacs.split.pbs][RNA-seq gene annotation & sequence discovery]] with multi-database mining in bash (Bassim 2016)
24. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.cross.species.slurm][Transcriptional cross-talk & gene networks]] in bash (Bassim 2016)
25. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.cross.species.R][Decorrelating network scores from RNA-seq]] dual RNA-seq data in R (Bassim 2016)
26. [[https://github.com/neocruiser/pipelines/blob/master/annotation/hmmscan-iacs.pbs][Gene annotation with Hidden Markov models]] using HMM profiles in bash (Bassim 2015)
27. [[https://github.com/neocruiser/pipelines/blob/master/calling/mappingV6.5.sh][Genomic filtering & variant calling with indel correction]] in bash (Bassim 2015)
28. [[https://github.com/neocruiser/pipelines/blob/master/calling/mapNoCount.sh][Genomic bam processing & variant calling QCs]] in bash (Bassim 2015)
29. [[https://github.com/neocruiser/thesis2014/blob/master/Paper3/paper3.R][Regularization & Ensemble learning]] on top of neural nets, GLMs & bagging in R (Bassim 2014)
30. [[https://github.com/neocruiser/thesis2014/blob/master/ebdbn/ebdbn%2520-%2520W2.R][Bayesian network inference from microarrays & timeseries data]] in R (Bassim 2013)
31. [[https://github.com/neocruiser/thesis2014/blob/master/mfuzz/mfuzz.R][Fuzzy clustering on Agilent microarray timeseries]] data for biomarker discovery in R (Bassim 2013)
32. [[https://github.com/neocruiser/thesis2014/blob/master/ordination/rda%2520-%2520W2.R][Ordination analyses of timeseries microarray data]] in R (Bassim 2013)
33. [[https://github.com/neocruiser/thesis2014/blob/master/microarrays/preProcessing_detailed.R][Agilent timeseries data processing and pattern extraction]] in R (Bassim 2012)
