#+TITLE: Translational Research: Pipelines in Data Mining and Systems Biology
#+AUTHOR: Sleiman Bassim, PhD
#+EMAIL: slei.bass@gmail.com

#+STARTUP: content
#+STARTUP: hidestars
#+OPTIONS: toc:5 H:5 num:3
#+LANGUAGE: english
#+LaTeX_HEADER: \usepackage[ttscale=.875]{libertine}
#+LATEX_HEADER: \usepackage[T1]{fontenc}
#+LaTeX_HEADER: \sectionfont{\normalfont\scshape}
#+LaTeX_HEADER: \subsectionfont{\normalfont\itshape}
#+LATEX_HEADER: \usepackage[innermargin=1.5cm,outermargin=1.25cm,vmargin=3cm]{geometry}
#+LATEX_HEADER: \linespread{1}
#+LATEX_HEADER: \setlength{\itemsep}{-30pt}
#+LATEX_HEADER: \setlength{\parskip}{0pt}
#+LATEX_HEADER: \setlength{\parsep}{-5pt}
#+LATEX_HEADER: \usepackage[hyperref]{xcolor}
#+LATEX_HEADER: \usepackage[colorlinks=true,urlcolor=SteelBlue4,linkcolor=Firebrick4]{hyperref}
#+EXPORT_SELECT_TAGS: export
#+EXPORT_EXCLUDE_TAGS: noexport

This git repository contains RNA-seq and microarray pipelines for transcriptome
assembly, gene expression, gene annotation, and other. Scripts dated
between 2015 and 2017 are running on [[https://www.xsede.org/][XSEDE]] supercomputers and [[http://www.iacs.stonybrook.edu/resources/handy-accounts#overlay-context=resources/accounts][LIred]] cluster in
the US, including Greenfield, Bridges, and LIred servers. All others
are running on H4H cluster at [[http://www.uhnresearch.ca/][UHN]].


Licence: MIT Licence

* Systemic Analytical Pipelines =scripts can be adjusted for any sample size=
1. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.h4h.pbs][Affymetrix sample preprocessing and automation]] in bash (Bassim 2018)
2. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.2.0.R][Affymetrix normalization and expression analyses]] in R (Bassim 2018)
3. [[https://github.com/neocruiser/pipelines/blob/master/r/affymetrix.summary.h4h.sh][Automation to create summary report of all expression analyses]] in
   bash (Bassim 2018)
4. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.h4h.pbs][Clustering and network automation]] in bash (Bassim 2018)
5. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.affymetrix.R][Network
   inference based on weighted scoring and different correlation and
   data transformation methods]] in R (Bassim 2018)
6. [[https://github.com/neocruiser/pipelines/blob/master/r/heatmaps.3.0.R][Clustering automation with different expression aggregation methods
   and bootstrapping]] in R (Bassim 2018)
7. [[https://github.com/neocruiser/pipelines/blob/master/r/classification.R][Machine learning to classify by bagging and ensemble learning for
   risk factor predictions]] in R (Bassim 2018)
8. [[https://github.com/neocruiser/pipelines/blob/master/debug/debug2.pbs][Separating reads from dual RNA sequencing of host and parasite]] in
   bash (Bassim 2017)
9. [[https://github.com/neocruiser/pipelines/blob/master/debug/debug4.slurm][Sequence abundance and expression from dual RNA-seq]] in R (Bassim 2017)
10. [[https://github.com/neocruiser/pipelines/blob/master/expression/filter-bridges.slurm][Reducing transcriptome size and noise by correcting for read
    abundance]] in bash (Bassim 2017)
11. [[https://github.com/neocruiser/pipelines/blob/master/mining/automated_analyses.sh][Aggregating and optimizing data structure for large scale
    annotation of cDNA sequences with over 13 databases, GO-terms,
    KEGG, reactome]] in bash (Bassim 2016) 
12. [[https://github.com/neocruiser/pipelines/blob/master/assembly/trinity-bridges.slurm][Transcriptome assembly of RNA-seq data optimized on high
    performance clusters]] in bash (Bassim 2016)
13. [[https://github.com/neocruiser/pipelines/blob/master/mapping/genome_guided_assemblies.pbs][Genome-guided assembly of RNA-seq data for new gene discovery]] in
    bash (Bassim 2016)
14. [[https://github.com/neocruiser/pipelines/blob/master/expression/degs-bridges.slurm][Differential
    gene expression for RNA-seq data using multi-method approach with
    summary report aggregation]] in bash and R (Bassim 2016)
15. [[https://github.com/neocruiser/pipelines/blob/master/annotation/kraken.db-bridges.slurm][Read classification for virus or most microbes identification from
    shotgun sequencing data]] in bash (bash 2016)
16. [[https://github.com/neocruiser/pipelines/blob/master/annotation/diamond-bridges.slurm][Fast sequence annotation against NCBI databases using blast and
    diamond]] in bash (Bassim 2016)
17. [[https://github.com/neocruiser/pipelines/blob/master/annotation/interproscan-bridges.slurm][Functional
    annotation of sequences by protein domains interrogating over 10
    databases]] in bash (Bassim 2016)
18. [[https://github.com/neocruiser/pipelines/blob/master/annotation/hmmscan-iacs.pbs][Functional annotation with Hidden Markov models using HMM profiles]]
    in bash (Bassim 2015)
19. [[https://github.com/neocruiser/pipelines/blob/master/annotation/blast-iacs.split.pbs][Sequence annotation of sequencing data against NCBI databases with
    more metadata per gene]] in bash (Bassim 2016)
20. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.cross.species.slurm][Creating genetic cross-talk networks from host and pathogen
    RNA-seq interactions]] in bash (Bassim 2016)
22. [[https://github.com/neocruiser/pipelines/blob/master/r/weighted.nets.cross.species.R][Calculating correlations and network scores from RNA-seq
    host-pathogen data]] in R (Bassim 2016)
23. [[https://github.com/neocruiser/pipelines/blob/master/calling/mappingV6.5.sh][Filtering and variant calling with indel correction on sequencing
    data]] in bash (Bassim 2015)
24. [[https://github.com/neocruiser/pipelines/blob/master/calling/mapNoCount.sh][Sorting, duplicate removal, and QC on variant calls]] in bash
    (Bassim 2015)
25. [[https://github.com/neocruiser/thesis2014/blob/master/ebdbn/ebdbn%2520-%2520W2.R][Bayesian network inference from Agilent micrarray timeseries data]]
    in R (Bassim 2013)
26. [[https://github.com/neocruiser/thesis2014/blob/master/Paper3/paper3.R][Regularization with lasso and ensemble learning build on top ANN,
    generalized linear model, and bagging]] in R (Bassim 2014)
27. [[https://github.com/neocruiser/thesis2014/blob/master/mfuzz/mfuzz.R][Fuzzy clustering on Agilent microarray timeseries data for
    biomarker discovery]] in R (Bassim 2013)
28. [[https://github.com/neocruiser/thesis2014/blob/master/ordination/rda%2520-%2520W2.R][Ordination analyses of timeseries microarray data]] in R (Bassim 2013)
29. [[https://github.com/neocruiser/thesis2014/blob/master/microarrays/preProcessing_detailed.R][Agilent timeseries data processing and pattern extraction]] in R
    (Bassim 2012)
