#!/bin/bash
#PBS -l nodes=1:ppn=16,walltime=168:00:00
#PBS -N string
#PBS -j oe
#PBS -q extended
#PBS -M sleiman.bassim@stonybrook.edu
#PBS -m abe
#PBS -V

# 1000 genes done in 1h w/ string
# 2000 genes done in 3h w/ NT

# DONT CHANGE #
###############
scratch=/gpfs/scratch/ballam
home=/gpfs/home/ballam
pbs=$(echo $PBS_JOBID | cut -f 1 -d '.')
nthreads=$(expr 2 \* 16)

# Files _CHANGE_
db=martia
blast=blastn
maxSeq=10

#p=4
#c=2
#file=DESeq2.raw.all.eXpress.tissue.p$p.c$c
#log=$scratch/ganglia/deg.raw.all/$file/diffExpr.P1e-${p}_C${c}.matrix.log2.dat
#assembled=$scratch/ganglia/assembled/raw.all.rscf.contigs.fa
# get gene ids and gene fasta sequences
#tmp=$project/tmp_$blast.$pbs
#mkdir $tmp
#cat $assembled | sed 's/.len*$//g' | perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' <(cat $log | cut -f1 | grep "^TRINITY" | sort - | uniq) - > $tmp/$file.contigs.$pbs.fa

query=$(find $scratch -maxdepth 4 -name "Trinit*GG.fa*" | egrep "para.*1")
output=$query.$db.$blast.$pbs.txt

# blast libraries
export PATH="$PATH:/gpfs/home/ballam/ncbi-blast-2.2.31+/bin"
export BLASTDB="/gpfs/scratch/ballam/db/$db"

## Full blast
time=$home/time
jobid=$file.$db
start=$(date); echo "Job started at: $start" > $time/$jobid.$pbs.time

cd $scratch/db/$db

$blast -query $query -db $db -outfmt " 7 qseqid qlen sseqid slen qstart qend sstart send evalue bitscore length pident nident mismatch gaps " -max_target_seqs $maxSeq -num_threads $nthreads -out $output

end=$(date); echo "Job ended at: $end" >> $time/$jobid.$pbs.time
