#!/bin/bash
#PBS -l nodes=1:ppn=16,walltime=24:00:00
#PBS -N ab
#PBS -j oe
#PBS -q long
#PBS -M sleiman.bassim@stonybrook.edu
#PBS -m abe
#PBS -V

# DONT CHANGE #
scratch=/gpfs/scratch/ballam
home=/gpfs/home/ballam
#ips=$scratch/db/ips/interproscan-5.18-57.0
ips=$scratch/db/ips/interproscan-5.16-55.0
pbs=$(echo $PBS_JOBID | cut -f 1 -d '.')

#peptides=$(find $scratch -maxdepth 4 -iname "nonmatching*prot.fa" | egrep "para.*1")
#output=$(dirname $peptides)
output=$scratch/nodule/clam-analysis/interpro/$pbs
tmp=$output/tmp_$pbs
mkdir -p $tmp

# Files _CHANGE_
p=3
c=2
matrix=ab
#file=edgeR.raw.all.eXpress.tissue.p$p.c$c
log=$scratch/nodule/clam-analysis/trinity_out_dir_40p/deg.eXpress.91*/edgeR*$matrix*/diffExpr.P1e-${p}_C${c}.$matrix.log2.dat
assembled=$scratch/nodule/clam-analysis/interpro/clam_40p_trinity.fa
contigs=$tmp/contigs.$matrix.$pbs.fa
peptides=$tmp/peptides.$matrix.$pbs.fa

time=$home/time
jobid=$pbs.ips
start=$(date); echo "Job started at: $start" > $time/$jobid.time
# Get gene ids (only differentially expressed)
# Get gene sequences
cat $assembled | sed 's/.len.*$//g' | perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' <(cat $log | cut -f1 | sed '1d' | sort - | uniq) - > $contigs
# Translate
transeq $contigs $peptides -frame=F --clean=yes
# Run interpro Scans
# Dont run HAMAP PRINTS
$ips/interproscan.sh -t p \
    -i $peptides \
    -appl ProDom,PANTHER,TIGRFAM,SUPERFAMILY,PIRSF,Gene3D,Pfam,SMART,PROSITEPROFILES,PROSITEPATTERNS,COILS \
    -iprlookup -goterms --pathways \
    -f TSV, HTML, SVG \
    --tempdir $tmp -d $output

rm -r $tmp
end=$(date); echo "Job ended at: $end" >> $time/$jobid.time

#    -f TSV, SVG, GFF3, XML, HTML \
#    -appl ProDom,PANTHER,TIGRFAM,SUPERFAMILY,PIRSF,Gene3D,Pfam,SMART,PROSITEPROFILES,PROSITEPATTERNS,COILS \
