#!/bin/bash
#PBS -l nodes=1:ppn=6,walltime=10:00:00,vmem=60g,mem=220g
#PBS -N affymetrix
#PBS -q himem
#PBS -V
#PBS -j oe
#PBS -m abe
#PBS -M slei.bass@gmail.com

set -x

#module load R\3.4.1
module load R

#### DO NOT TOUCH
# cluster directories
# server characteristics and files indexing
scratch=/cluster/projects/kridelgroup
home=/cluster/home/sbassim

pbs=$(echo $PBS_JOBID | cut -f1 -d '.')
project=$scratch/relapse
output=$project/expression/$pbs

# time monitoring
time=$home/time
jobid=$pbs.R.affymetrix
start=$(date); echo "Job started at: $start" > $time/$jobid.time

## Directory restructuring
mkdir -p $output/figures $output/summary $output/scripts
cd $output


#### USER DEFINED
## subset dataset and remove ncRNA probes
discard=yes



############
## PART I ##
############
## listing the selected samples for analysis
## phenodata is a mantained sample classification file
## depricated
#ls -trlgG $project/raw/*CEL | awk '{print $7}' | cut -f4 -d '_' | sed 's/.CEL//g' | sort - > $output/summary/sampleIDs
cp $project/summary/sampleIDs $output/summary
cp $project/summary/phenodata $output/summary
cp $project/summary/abc_gcb.genes.counts $output/summary
cp $project/summary/annotated.normalized.systemic.expression.array.txt $output/summary

## Discard probes related to RNAs or just ncRNA
## they account for 46.68% of the array.
## Array (75.523 probes) among which 35.253 annotated ncRNA
## sorting and removing duplicated rows corrects for NAs from array construction
if [ $discard == "yes" ]; then

    ## get list of RNA mentions
    touch  $output/patterns.flagged4removal.txt
    for pat in rna pseudo uncharac intron; do
	grep -iwo "\w*${pat}\w*" $output/summary/annotated.normalized.systemic.expression.array.txt | \
	    sort - | uniq -c | sort -k1 -rn >> $output/patterns.flagged4removal.txt
    done

    ## remove all non coding RNAs, pseudogenes, uncharacterized, NAs, and intronic
    ## patterns were "manually" flagged for removal using the previous filtering output
    ## keep RNA and mRNA mentions
    ## uniq -u will remove NAs with no probe IDs (around 8k from 75k)
    ## additional step will remove 24k NA with probe IDs but without functional annotations
    ## sed '1d' will remove the header of the annoation file (else error matrix generation inside R)
    ## antisense genes left in data (around 7k)
    ## novel transcripts left in data (around 12k)

    ## remove NAs (everything)
    grep -Fwvf <(cat $output/summary/annotated.normalized.systemic.expression.array.txt | sed 's/ /./g' | \
			cut -f1,8 | grep -w "NA" | cut -f1 | sort - | uniq -u) \
	 $output/summary/annotated.normalized.systemic.expression.array.txt > \
	 $output/summary/annotated.normalized.systemic.expression.woNAs.txt
    
    grep -Fivf <(cat $project/summary/patterns.flagged4removal_final.txt | \
			awk '{print $2}') $output/summary/annotated.normalized.systemic.expression.woNAs.txt | \
	sed '1d' | cut -f1 | sort - | uniq -u > $output/ids.wo.ncrna
    
    touch subsetting.ncrna.OK
fi


## The HTA file was released in 2016 by affymetrix
## It contains probeset_ids and transcript_clusterIDs and gene annotations
## In order to make use of it, the RMA should be initiated with probeset rather than the core argument
## cat $project/summary/HTA-2_0.na36.hg19.probeset.csv | sed -e '1,14d' | sed -e 's/"//g' -e 's/,/\t/g'  | cut -f1,7,11 | awk '{if($3 != "---")print$0}' | sed -e 's/hg\/.*\t/hg\t/g' | sed 's/.\/\/./"/g' | cut -f1,2 -d'"' | sed 's/hg\t.*"/hg\t/g' | less


## Indexing, normalization, annoation, differential expression, and plotting of affymetrix data
R CMD BATCH $home/script*/r/affymetrix.2.0.R
touch gene.expression.OK



#############
## PART II ##
#############
## summarizing the output of differential genes
cd $output
mv $output/*txt $output/summary

#rm -r $output/ffObjs
cp $home/script*/r/affymetrix* $output/scripts


# Choose expression parameters: B-stats, adjusted Pval, average expression, logFC
## -7.358 is the minimum Bstatistics of diff genes
bstatistics=-8
adjustedpvalue=.1
averageexpression=2
highfoldchange=.5
lowfoldchange=-.5

bash $home/script*/r/affymetrix.summary.h4h.sh $bstatistics $adjustedpvalue $averageexpression $highfoldchange $lowfoldchange $pbs $output
sleep 100


## -4.6 is the Bstats that contains the third quantile of the distribution of differential genes (-4.12 is the mean)
bstatistics=-4
bash $home/script*/r/affymetrix.extract.genes.sh $bstatistics $output
sleep 100



##############
## PART III ##
##############
## clean up and organization of the output repo
cd $output
mv $output/*pdf $output/figures
rm $output/.RData

tar cf $output/$pbs.affymetrix.figures.tar $output/figures
gzip $output/$pbs.affymetrix.figures.tar

tar cf $output/$pbs.affymetrix.summary.tar $output/summary
gzip $output/$pbs.affymetrix.summary.tar

end=$(date); echo "Job ended at: $end" >> $time/$jobid.time

